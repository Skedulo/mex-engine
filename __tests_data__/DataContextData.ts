import {PageLevelDataContext} from "@skedulo/mex-engine-proxy";

export const SimpleProductData1 = (): PageLevelDataContext => {
    let newObjectKey = "temp-xxx";

    return {
        formData: {
            Products:
                [
                    {UID: "1", __typename: "Products"},
                    {UID: "2", __typename: "Products"}
                ]
        },
        pageData: {
            UID: newObjectKey,
            __typename: "Products",
            __isTempObject: true,
            title: "My Product",
            qty: 2,
            createdDate: "2022-01-01",
            userInfo: {
                Name: "Huy Vu",
                email: 'hieu.bui@skedulo.com',
                url: 'https://www.google.com.vn/'
            },
            users: [],
        },
        sharedData: {},
        metadata: {
            contextObjectId: "temp-object-id",
            packageId: "temp-package"
        }
    }
}

export const MockSearchDataContext = () => {
    return ({
            formData: {
                JobProducts:
                    [
                        {
                            CreatedDate: "2022-05-31T01:04:18.000Z",
                            Product: {
                                Description: "Power sore used for cutting wood with efficiency.",
                                Name: "Power Saw",
                                UID: "01t3t00000409jbAAA",
                                __typename: "Products",
                            },
                            ProductId: "01t3t00000409jbAAA",
                            ProductName: "Power Saw",
                            Qty: 67,
                            UID: "a0K3t00000Z4wfiEAB",
                            __typename: "JobProducts",
                        }, {
                        CreatedDate: "2022-05-31T01:04:18.000Z",
                        Product: {
                            Description: "Power sore used for cutting wood with efficiency.",
                            Name: "Power Saw",
                            UID: "01t3t00000409jbAAA",
                            __typename: "Products",
                        },
                        ProductId: "01t3t00000409j1AAA",
                        ProductName: "Power Saw",
                        Qty: 67,
                        UID: "a0K3t00000Z4wfiEAB",
                        __typename: "JobProducts",
                    }
                    ]
            },
            pageData: {},
            sharedData: {},
            metadata: {}
        }
    )
}
