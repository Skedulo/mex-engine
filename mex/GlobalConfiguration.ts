'use strict'

export enum DevelopmentModeConfigEnum {
    UseLocalInstanceData = "UseLocalInstanceData",
    UseLocalStaticData = "UseLocalStaticData",
    UseLocalUIDefinition = "UseLocalUIDefinition",
    UseLocalTranslationFiles = "UseLocalTranslationFiles",
    UseLocalCustomFunctions = "UseLocalCustomFunctions",
    UseLocalFormMetadata = "UseLocalFormMetadata",
}

export type DevelopmentFormConfig = {
    FormData: () => any,
    StaticData?: () => any,
    UiDef: () => any,
    Languages: [{ tag: string, content: () => any}],
    FormMetadata: () => any,
}

class GlobalConfig {

    EnableDevelopment: boolean =  __DEV__
    FormName: string = "local"

    DevelopmentFormConfig: Map<string, DevelopmentFormConfig> = new Map<string, DevelopmentFormConfig>()

    constructor() {
        this.DevelopmentFormConfig.set("local", {
            FormData: () => require("../local_data/instance_data.json"),
            StaticData: () => require("../local_data/static_data.json"),
            UiDef: () => require("../local_data/ui_def.json"),
            Languages: [{
                tag: "en",
                content: () => require("../local_data/resources/locale/en.json")
            }],
            FormMetadata: () => require("../local_data/metadata.json"),
        })

        // Uncomment these to reduce package size, otherwise these files will be included in the engine
        // this.DevelopmentFormConfig.set("showcase", {
        //     FormData: () => require("../mex-test-forms/showcase/fake_data/instance_data.json"),
        //     StaticData: () => require("../mex-test-forms/showcase/fake_data/instance_data.json"),
        //     UiDef: () => require("../mex-test-forms/showcase/mex_definition/ui_def.json"),
        //     Languages: [{
        //         tag: "en",
        //         content: () => require("../mex-test-forms/showcase/mex_definition/static_resources/locales/en.json")
        //     }]
        // })
        //
        // this.DevelopmentFormConfig.set("leave", {
        //     FormData: () => require("../mex-test-forms/leave_management/fake_data/instance_data.json"),
        //     StaticData: () => require("../mex-test-forms/leave_management/fake_data/static_data.json"),
        //     UiDef: () => require("../mex-test-forms/leave_management/mex_definition/ui_def.json"),
        //     Languages: [{
        //         tag: "en",
        //         content: () => require("../mex-test-forms/leave_management/mex_definition/static_resources/locales/en.json")
        //     }]
        // })
        //
        // this.DevelopmentFormConfig.set("lone", {
        //     FormData: () => require("../mex-test-forms/lone_worker_checkin/fake_data/instance_data.json"),
        //     StaticData: () => require("../mex-test-forms/lone_worker_checkin/fake_data/static_data.json"),
        //     UiDef: () => require("../mex-test-forms/lone_worker_checkin/mex_definition/ui_def.json"),
        //     Languages: [{
        //         tag: "en",
        //         content: () => require("../mex-test-forms/lone_worker_checkin/mex_definition/static_resources/locales/en.json")
        //     }]
        // })
        //
        // this.DevelopmentFormConfig.set("foodreview", {
        //     FormData: () => require("../mex-test-forms/food_review/fake_data/instance_data.json"),
        //     StaticData: () => require("../mex-test-forms/food_review/fake_data/static_data.json"),
        //     UiDef: () => require("../mex-test-forms/food_review/mex_definition/ui_def.json"),
        //     Languages: [{
        //         tag: "en",
        //         content: () => require("../mex-test-forms/food_review/mex_definition/static_resources/locales/en.json")
        //     }]
        // })
    }

    private DevelopmentModeConfig = {
        [DevelopmentModeConfigEnum.UseLocalInstanceData]: false,
        [DevelopmentModeConfigEnum.UseLocalStaticData]: false,
        [DevelopmentModeConfigEnum.UseLocalUIDefinition]: true,
        [DevelopmentModeConfigEnum.UseLocalTranslationFiles]: false,
        [DevelopmentModeConfigEnum.UseLocalCustomFunctions]: false,
        [DevelopmentModeConfigEnum.UseLocalFormMetadata]: false
    }

    canUseDevelopmentConfig(config: DevelopmentModeConfigEnum) {
        return this.EnableDevelopment && this.DevelopmentModeConfig[config]
    }

    getDevelopmentFormConfig(): DevelopmentFormConfig {
        return this.DevelopmentFormConfig.get(this.FormName)!
    }
}

export default new GlobalConfig() as GlobalConfig
