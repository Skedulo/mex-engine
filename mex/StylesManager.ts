import ThemeManager from './colors/ThemeManager'
import {Platform} from "react-native";

type StyleConstant = {
    defaultHorizontalPadding: number
    defaultVerticalPadding: number,
    smallVerticalPadding: number,
    componentVerticalPadding: number,
    betweenTextSpacing: number,
    captionTextSize: number,
    betweenComponentVerticalSpacing: number
}

class StylesManager {
    styles: any;
    stylesConstant: StyleConstant = {
        defaultHorizontalPadding: 16,
        defaultVerticalPadding: 16,
        smallVerticalPadding: 8,
        componentVerticalPadding: 12,
        betweenTextSpacing: 6,
        captionTextSize: 12,
        betweenComponentVerticalSpacing: 32
    };

    initializeStyles() {
        let colors = ThemeManager.getColorSet()

       this.styles =  {
           text: {
               "color": colors.skeduloText,
               "fontSize": 20,
           },

           errorText: {
               "color": colors.red800,
               "fontSize": 14,
           },
           textRegular: {
               fontWeight: "normal",
               fontSize: 16,
               flexWrap: 'wrap',
               color: colors.skeduloText
           },
           textTitleListItem: {
               fontWeight: "600",
               fontSize: 16,
               color: colors.skeduloText
           },
           textCaptionListItem: {
               fontWeight: "normal",
               fontSize: 16,
               color: colors.navy600
           },
           textCaptionSelector: {
               fontWeight: "normal",
               fontSize: 14,
               color: colors.navy600
           },
           textCaption: {
               fontWeight: "normal",
               fontSize: 12,
               flexWrap: 'wrap',
               color: colors.skeduloText
           },
           textMedium: {
               fontWeight: "500",
               fontSize: 16,
               color: colors.skeduloText
           },

           textHeadingBold: {
               fontWeight: "bold",
               fontSize: 18,
               color: colors.skeduloText
           },

           navigationTitle: {
               fontSize: 16,
               fontWeight: "bold"
           },

           navigationSubtitle: {
               fontSize: 14
           },
           editText: {
               borderColor: colors.navy300,
               borderWidth: 1,
               borderRadius: 2,
               marginTop: 8,
               height: 48,
               fontSize: 16,
               paddingHorizontal: 16,
               justifyContent: 'center',
               color: colors.skeduloText,
           },
           selector: {
               borderWidth: 1,
               borderRadius: 2,
               marginTop: 8,
               height: 48,
               paddingHorizontal: 16,
           },
           searchBarTextContainer: {
               borderColor: colors.navy100,
               backgroundColor: colors.navy75,
               borderWidth: 1,
               borderRadius: 10,
               paddingHorizontal: 12,
           },
           searchBarText: {
               backgroundColor: "transparent",
               borderWidth: 0,
               borderRadius: 0,
               paddingHorizontal: 12,
               paddingTop: 12,
               paddingBottom: 12,
               color: colors.skeduloText
           },
           button: {
               backgroundColor: colors.skeduloBlue,
               height: 44,
               borderRadius: 8,
               paddingVertical: 10,
               paddingHorizontal: 12,
               justifyContent: "center"
           },
           buttonSmall: {
               backgroundColor: colors.skeduloBlue,
               borderRadius: 8,
               justifyContent: "center",
               height: 32,
               paddingVertical: 8,
               paddingHorizontal: 20
           },
           buttonDefault: {
               height: 44,
               borderRadius: 8,
               paddingVertical: 10,
               paddingHorizontal: 12,
               justifyContent: "center",
               borderColor: colors.skeduloBlue,
               borderWidth: 1
           },
           headerButtonText: {
               fontSize: 16,
               color: colors.white,
               fontWeight: 'normal',
               alignSelf: 'center',
           },
           buttonInsideText: {
               fontSize: 16,
               color: colors.white,
               fontWeight: "600",
               alignSelf: "center",
           },
           buttonSmallInsideText: {
               fontSize: 14,
               color: colors.white,
               fontWeight: "600",
               alignSelf: "center",
           },
           buttonDefaultInsideText: {
               fontSize: 16,
               color: colors.skeduloBlue,
               fontWeight: "500",
               alignSelf: "center",
           },

           buttonLinkText: {
               fontSize: 16,
               color: colors.skeduloText,
               fontWeight: "500",
               alignSelf: "center",
           },

           headerButton: {
               color: colors.white
           },
           headerButtonAlt: {
               color: colors.skedBlue800
           },
           checkBoxProps: {
               size: 25,
               useNativeDriver: true,
               unfillColor: colors.white,
               disableBuiltInState: true,
               iconStyle: {
                   borderRadius: Platform.OS === 'ios' ? 13 : 5
               }
           },
           headerBarSelectView: {
               borderRadius: 8,
               flexDirection: 'row',
               justifyContent:'space-between',
               paddingTop: 19,
               paddingBottom:15,
               paddingHorizontal: 16
           }
       }
    }

    getStyles() { return this.styles }
    getStyleConst() : StyleConstant { return this.stylesConstant }

}

export default new StylesManager() as StylesManager

