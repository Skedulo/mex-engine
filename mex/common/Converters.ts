import {translate} from "../assets/LocalizationManager";
import {NativeModules} from 'react-native';
import {BaseStructureObject} from "./models/BaseStructureObject";
import moment from "moment";
import AssetsManager from "../assets/AssetsManager";
import {TimeZoneType} from "@skedulo/mex-engine-proxy/dist/src/proxies/utils/models";

let cache = require('js-cache')();

let converters = {
    date: {
        dateFormat: function (value: string , timezoneType?: TimeZoneType): Promise<string|undefined> {
            return getCachedValueOrFromNative(value, "date", timezoneType)
        },

        timeFormat: function (value: string , timezoneType?: TimeZoneType) {
            if (value && value != '') {
                let timeValues = value.split(":")
                let dashes = value.split("-")

                if (dashes.length == 1 && timeValues.length == 3) {
                    let [hour, minutes, seconds] = value.split(":")

                    // Remove ticks in case of "Time"
                    let dotIndex = seconds.indexOf(".")

                    if (dotIndex !== -1) {
                        seconds = seconds.slice(0, dotIndex)
                        value = [hour, minutes, seconds].join(":")
                    }
                }
            }

            return getCachedValueOrFromNative(value, "time", timezoneType)
        },

        dateTimeFormat: function(value: string, timezoneType?: TimeZoneType) {
            return getCachedValueOrFromNative(value, "datetime", timezoneType)
        },

        dateTimeRangeFormat: function(startDateTime:string, endDateTime:string, timezoneType?: TimeZoneType) {
            let timezone = typeof timezoneType == "string" ? timezoneType :  "none"

            if (!startDateTime || !endDateTime) {
                return undefined;
            }
            let jobId = timezone === "job"
                ? AssetsManager.cachedContextId ?? undefined
                : undefined

            return NativeModules.MexMainModule.getLocalizedDateTimeRangeDisplay(startDateTime, endDateTime, timezoneType, jobId)
        },

        durationFormat: function(d1: string, d2: string, _: string): Promise<string> {
            // TODO: This should be implemented from Native side, by another task https://skedulo.atlassian.net/browse/ENG-33287
            let datetime1 = moment(d1)
            let datetime2 = moment(d2)
            let duration = moment.duration(datetime2.diff(datetime1))

            let result = (duration.days() * 24 + duration.hours()) + " hours and " + duration.minutes() + ' minutes'

            return Promise.resolve(result);
        }
    },
    localization: {
        translate: function(key:string): string {
            return translate(key)
        }
    },
    data: {
        isTempUID: function(obj: BaseStructureObject) {
            return obj.__isTempObject ? true : false
        },
        isCreatingNewObject: function(obj: BaseStructureObject) {
            return obj.__isCreatingNewObject ? true : false
        }
    }
}

function getCachedValueOrFromNative(value: string, type: string, timezoneType?: TimeZoneType): Promise<string|undefined> {
    let timezone = typeof timezoneType == "string" ? timezoneType :  "none"
    if (!value) {
        return Promise.resolve(undefined)
    }

    let cacheKey = `converters-datetimeFormat-${value}-${type}-${timezoneType}`;

    let cacheValue = cache.get(cacheKey, null);
    if (cacheValue) {
        return Promise.resolve(cacheValue)
    }
    let jobId = timezone === "job"
        ? AssetsManager.cachedContextId ?? undefined
        : undefined

    let promise = new Promise<string|undefined>(function (resolve, reject) {
        NativeModules.MexMainModule.getLocalizedDateTimeDisplay(value, type, timezone, jobId)
            .then((value:string) => {
                cache.set(cacheKey, value, 30000); // 30 secs

                resolve(value)
            })
            .catch((err:Error) => {
                reject(err)
            })
    })

    cache.set(cacheKey, promise, 30000); // 30 secs

    return promise
}

export default converters
