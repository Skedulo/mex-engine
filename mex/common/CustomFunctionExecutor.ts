import AssetsManager from "../assets/AssetsManager";

export default class CustomFunctionExecutor {

    functionMap: Map<string, (...args: any[]) => any> = new Map<string,  (...args: any[]) => any>()

    setup(functionName:string): void {
        let parts = functionName.split(".")

        // get only last part
        functionName = parts[parts.length-1]

        // @ts-ignore
        if (this.functionMap.has(functionName)) {
            return
        }

        this.functionMap.set(functionName, this.buildCustomFunction(functionName))
    }

    returnMappingObject(): any {
        let object:any = {}

        Array.from(this.functionMap.keys()).forEach(key => {
            object[key] = this.functionMap.get(key)
        })

        return object;
    }


    private buildCustomFunction(functionName: string) : (...args: any[]) => any {
        let customJs = AssetsManager.getCustomFunction();

        if (!customJs || customJs === "") {
            // Custom function does not exist, proceed
            throw new Error("Not found custom function")
        }

        return (...args: any[]) => {
            let argsArray: any[] = []
            let argsContext: any = {}

            for (let i = 0; i < args.length; i++) {
                let argName = "arg_" + i

                argsArray.push(argName)

                argsContext[argName] = args[i]
            }

            let argsStr = argsArray.join(', ')

            customJs += "\n" + `exportFns.${functionName}(${argsStr})`

            let result = this.evalInContext(customJs!, argsContext)

            return result;
        }
    }

    // @ts-ignore
    private evalInContext(js:string, context:any) {
        return eval('with(context) { ' + js + ' }');
    }

}
