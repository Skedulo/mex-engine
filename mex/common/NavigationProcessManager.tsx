import * as RootNavigation from "./RootNavigation";
import {Routing, RoutingPageDef, RoutingType} from "@skedulo/mex-types";
import Expressions from "./expression/Expressions";
import {translateOneLevelOfExpression} from "./InternalUtils";
import {INavigationProcessManager, PageLevelDataContext} from "@skedulo/mex-engine-proxy";

export class NavigationContext {
    constructor() { }

    // This variable behaves as an additional instruction for upsert for letting the Form know where to add to the object to for the prevPageNavigationContext
    sourceExpression: string|undefined

    prevPageNavigationContext: NavigationContext|undefined

    currentDataContext:PageLevelDataContext|undefined
}

class NavigationProcessManager implements INavigationProcessManager {
    navigationMap = new Map();

    /**
     *
     * @param pageId
     * @return Promise
     */
    addTrack(pageId: string): Promise<any> {
        return new Promise((resolve, reject) => {
            this.navigationMap.set(pageId, [resolve, reject])
        })
    }

    resolveTrack(pageId:string, value:any) {
        if (!this.navigationMap.has(pageId)) {
            return;
        }

        let [resolve, _] = this.navigationMap.get(pageId)

        resolve(value)

        this.navigationMap.delete(pageId)
    }

    async navigate(routing: RoutingType, pageData: any, navigationContext: NavigationContext|undefined = undefined, currentDataContext: any|undefined = undefined): Promise<any> {
        if (typeof(routing) === 'string' || routing instanceof String) {
            // Normal flow, just define as a string
            let pageId = routing as string

            RootNavigation.navigate(pageId, {pageData: pageData, navigationContext: navigationContext})

            return await this.addTrack(pageId)

        } else {
            // Normal flow, just define as a string
            routing = routing as RoutingPageDef

            let newDataContext = { ...currentDataContext }

            if (routing.ui) {
                // If UI is defined, start opening middleware page for allowing provide more data
                RootNavigation.navigate("routingModalScreen", {dataContext: currentDataContext, jsonDef: routing, navigationContext: navigationContext})

                let routingData = await this.addTrack("routingModalScreen")

                if (!routingData) {
                    // Do nothing, the user seem to cancel
                    return
                }

                newDataContext.routing = routingData
            }

            let route:Routing|undefined = undefined

            for (let i = 0; i < routing.routing.length; i++) {
                let eachRoute = routing.routing[i]

                if (!eachRoute.condition) {
                    route = eachRoute
                    break;
                }

                let isTrue = Expressions.getValueExpression({ expressionStr: eachRoute.condition, dataContext: newDataContext})

                if (isTrue) {
                    route = eachRoute
                    break;
                }
            }

            if (!route) {
                return;
            }

            if (pageData === undefined) {
                if (route.transferData) {
                    // We want to get the object directly, we don't want to flat it out in case user just want to directly pass entire pageData from prev page to this
                    pageData = translateOneLevelOfExpression({jsonDef: route.transferData, dataContext: newDataContext});
                }
            } else {
                pageData = {
                    ...pageData,
                    ...(!route.transferData ? {} : translateOneLevelOfExpression({jsonDef: route.transferData, dataContext: newDataContext}))
                }
            }

            RootNavigation.navigate(route.page, {pageData: pageData, navigationContext: navigationContext})

            return await this.addTrack(route.page)
        }
    }

    goBack(output?: any|undefined): void {

        let currentRouteName = RootNavigation.getCurrentRouteName();

        RootNavigation.goBack()

        this.resolveTrack(currentRouteName, output)
    }

}

export default new NavigationProcessManager() as NavigationProcessManager
