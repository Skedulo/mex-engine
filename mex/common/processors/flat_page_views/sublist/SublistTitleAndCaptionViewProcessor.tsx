import * as React from "react";
import {Text} from "react-native";
import {View} from "react-native";
import AbstractSublistViewProcessor, {SublistPageViewArgs} from "./AbstractSublistViewProcessor";
import StylesManager from "../../../../StylesManager";
import Expressions from "../../../expression/Expressions";
import SkedStaticResourceIcon from "../../../../../components/SkedStaticResourceIcon";
import MexAsyncText from "../../../../../components/MexAsyncText";
import SkedIcon from "../../../../../components/SkedIcon";
import {SublistTitleAndCaptionViewComponentModel} from "@skedulo/mex-types";
import {IconTypes} from "@skedulo/mex-engine-proxy";

export default class SublistTitleAndCaptionViewProcessor extends AbstractSublistViewProcessor<SublistTitleAndCaptionViewComponentModel> {

    getTypeName(): string {
        return "subListTitleAndCaption";
    }

    generateInnerSublistViewComponent(args: SublistPageViewArgs<SublistTitleAndCaptionViewComponentModel>): JSX.Element {

        let styles = StylesManager.getStyles()
        let styleConsts = StylesManager.getStyleConst()

        const getTitle = async (): Promise<string> => {
            if (!args.jsonDef.title || args.jsonDef.title === "")
                return ''

            let title = Expressions.getValueFromLocalizedKey({expressionStr: args.jsonDef.title, dataContext: args.dataContext})

            if (title instanceof Promise) {
                return await title
            }

            return title;
        }

        const getCaption = async (): Promise<string> => {
            let caption = Expressions.getValueFromLocalizedKey({expressionStr: args.jsonDef.caption, dataContext: args.dataContext})

            if (caption instanceof Promise) {
                return await caption
            }

            return caption;
        }

        let image = args.jsonDef.image

        return (
            <View style={{ flexDirection: "row"}}>
                {image
                    ? <SkedStaticResourceIcon
                        icon={image.name}
                        style={{
                            marginRight: styleConsts.betweenTextSpacing,
                            height: 40,
                            width: 40
                        }} />
                    : null}

                <View style={{flex: 1}}>
                    {args.jsonDef.title && args.jsonDef.title !== "" ? (<MexAsyncText promiseFn={getTitle}>
                        {(text) => (
                            <Text style={styles.textRegular}>{text}</Text>)}
                    </MexAsyncText>): null }

                    <MexAsyncText promiseFn={getCaption}>
                        {(text) => (
                            <Text style={[styles.textRegular, { marginTop: 5}]}>
                                {text}
                            </Text>)}
                    </MexAsyncText>
                </View>

                <SkedIcon style={{
                    alignSelf: "center",
                    height: 10,
                    width: 6
                }} iconType={IconTypes.ChevronRight}/>
            </View>);
    }

    override useObservable() {
        return true;
    }
}
