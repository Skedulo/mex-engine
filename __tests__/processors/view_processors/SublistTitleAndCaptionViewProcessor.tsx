import React from 'react';
import {act, render} from "@testing-library/react-native";
import {DevelopmentModeConfigEnum} from "../../../mex/GlobalConfiguration";
import {
    SublistPageViewArgs
} from "../../../mex/common/processors/flat_page_views/sublist/AbstractSublistViewProcessor";
import {SimpleProductData1} from "../../../__tests_data__/DataContextData";
import {
    SublistTitleAndCaptionViewComponentModel
} from "@skedulo/mex-types";
import {NavigationContext} from "../../../mex/common/NavigationProcessManager";
import StylesManager from "../../../mex/StylesManager";
import SublistTitleAndCaptionViewProcessor
    from "../../../mex/common/processors/flat_page_views/sublist/SublistTitleAndCaptionViewProcessor";
import Localization from "../../../mex/assets/LocalizationManager";

jest.mock("../../../mex/GlobalConfiguration", () => {
    const originalModule = jest.requireActual("../../../mex/GlobalConfiguration");

    originalModule.default.canUseDevelopmentConfig = (_: DevelopmentModeConfigEnum) => true

    return originalModule
})

jest.mock("../../../mex/assets/LocalizationManager", () => {
    let mock:any;

    mock = jest.requireActual("../../../mex/assets/LocalizationManager")

    mock.default.loadFromLocalResources = () => {
        return [
            ["en"],
            {
                "en": {
                    "title": "Title",
                    "caption": "Caption"
                }
            }
        ]
    }

    return mock
})

beforeEach(() => {
    jest.resetModules()
        .resetAllMocks()
})



describe('Sublist title and caption view processor tests', function () {


    it('SublistTitleAndCaptionViewProcessor basic render', async function () {

        jest.setTimeout(10_000)

        await Localization.initializeLocalization();
        StylesManager.initializeStyles()

        let viewProcessor = new SublistTitleAndCaptionViewProcessor();

        let argsModel: SublistTitleAndCaptionViewComponentModel = {
            title: "title",
            caption: "caption",
            type: "subListTitleAndCaption"
        }

        let args:SublistPageViewArgs<SublistTitleAndCaptionViewComponentModel> = {
            dataContext: SimpleProductData1(),
            jsonDef: argsModel,
            navigationContext: new NavigationContext()
        };

        let Comp = viewProcessor.generateComponent()

        let renderer = render(<Comp args={args} />)

        expect(renderer.toJSON()).toMatchSnapshot()

        await act(async () => {
            await new Promise((r) => setTimeout(r, 20));
        });

        expect(renderer.toJSON()).toMatchSnapshot()
    });

});
