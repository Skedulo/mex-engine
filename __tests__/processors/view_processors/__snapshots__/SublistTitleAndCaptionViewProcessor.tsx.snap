// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sublist title and caption view processor tests SublistTitleAndCaptionViewProcessor basic render 1`] = `
<View
  style={
    Object {
      "alignContent": "flex-start",
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#223049",
            "flexWrap": "wrap",
            "fontSize": 16,
            "fontWeight": "normal",
          }
        }
      />
      <Text
        style={
          Array [
            Object {
              "color": "#223049",
              "flexWrap": "wrap",
              "fontSize": 16,
              "fontWeight": "normal",
            },
            Object {
              "marginTop": 5,
            },
          ]
        }
      />
    </View>
    <Image
      resizeMode="contain"
      source={
        Object {
          "uri": "ChevronRight",
        }
      }
      style={
        Object {
          "alignSelf": "center",
          "height": 10,
          "width": 6,
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#E5E7EE",
        "height": 1,
        "marginTop": 6,
        "width": "100%",
      }
    }
  />
</View>
`;

exports[`Sublist title and caption view processor tests SublistTitleAndCaptionViewProcessor basic render 2`] = `
<View
  style={
    Object {
      "alignContent": "flex-start",
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#223049",
            "flexWrap": "wrap",
            "fontSize": 16,
            "fontWeight": "normal",
          }
        }
      >
        Title
      </Text>
      <Text
        style={
          Array [
            Object {
              "color": "#223049",
              "flexWrap": "wrap",
              "fontSize": 16,
              "fontWeight": "normal",
            },
            Object {
              "marginTop": 5,
            },
          ]
        }
      >
        Caption
      </Text>
    </View>
    <Image
      resizeMode="contain"
      source={
        Object {
          "uri": "ChevronRight",
        }
      }
      style={
        Object {
          "alignSelf": "center",
          "height": 10,
          "width": 6,
        }
      }
    />
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#E5E7EE",
        "height": 1,
        "marginTop": 6,
        "width": "100%",
      }
    }
  />
</View>
`;
